lms/events/__init__.py

```diff
@@ -0,0 +1,107 @@
+import json
+from dataclasses import dataclass, field, fields
+from typing import List, Optional
+
+from pyramid.events import subscriber
+from pyramid.request import Request
+
+from lms.models import EventType
+from lms.services import EventService, LTIRoleService
+
+
+@dataclass
+class BaseEvent:  # pylint:disable=too-many-instance-attributes
+    """Base class for generic events."""
+
+    request: Request
+    """Reference to the current request"""
+
+    type: EventType.Type
+    """Type of the event"""
+
+    user_id: Optional[int] = None
+    """Which user is related to this event"""
+
+    role_ids: List[int] = field(default_factory=list)
+    """Which roles does the user have in relation to this event"""
+
+    application_instance_id: Optional[int] = None
+    course_id: Optional[int] = None
+    assignment_id: Optional[int] = None
+    grouping_id: Optional[int] = None
+
+    data: Optional[dict] = None
+    """Extra data to associate with this event"""
+
+    def __post_init__(self):
+        for event_field in fields(self):
+            getter_name = f"_get_{event_field.name}"
+
+            # If we don't have a value for `field` and we have implemented
+            # a _get_`field.name` method, use that to get a default
```

Comment:
This is a bit of a soup of gettattr/hasattr/setattr but avoids a lot of `if not self.field` below.

All those conditionals were making pyling angry and it was getting a bit hard to read.